File: backend/src/test/java/cz/forgottenempire/servermanager/steamauth/SteamAuthServiceUnitTest.java
package cz.forgottenempire.servermanager.steamauth;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import cz.forgottenempire.servermanager.steamcmd.SteamCmdAuthService;
import cz.forgottenempire.servermanager.workshop.SteamAuthDto;

import java.util.Collections;
import java.util.List;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class SteamAuthServiceUnitTest {

    @Mock
    private SteamAuthRepository steamAuthRepository;

    private SteamAuthService service;

    @BeforeEach
    void setUp() {
        service = new SteamAuthService(steamAuthRepository);
    }

    @Test
    void whenGetAuthAccountAndSteamAuthSetup_thenReturnCorrectSteamAuth() {
        SteamAuth steamAuth = new SteamAuth(1L, "username", "hunter2", "ABCDE");
        when(steamAuthRepository.findAll()).thenReturn(List.of(steamAuth));

        SteamAuth authAccount = service.getAuthAccount();

        assertThat(authAccount).isEqualTo(steamAuth);
    }

    @Test
    void whenGetAuthAccountAndNoSteamAuthSetup_thenReturnEmptySteamAuth() {
        when(steamAuthRepository.findAll()).thenReturn(Collections.emptyList());

        SteamAuth authAccount = service.getAuthAccount();

        assertThat(authAccount).isEqualTo(new SteamAuth());
    }

    @Test
    void whenSetSteamAccountWithNewPassword_thenPasswordIsUpdated() {
        SteamAuth prevAuth = new SteamAuth(1L, "username", "hunter2", "ABCDE");
        when(steamAuthRepository.findAll()).thenReturn(List.of(prevAuth));
        SteamAuthDto newAuth = new SteamAuthDto();
        newAuth.setUsername("new_username");
        newAuth.setPassword("n3wp4ssw0rd");
        newAuth.setSteamGuardToken("FGHIJ");

        service.setAuthAccount(newAuth);

        SteamAuth actualAuthAccount = service.getAuthAccount();
        assertThat(actualAuthAccount.getUsername()).isEqualTo("new_username");
        assertThat(actualAuthAccount.getPassword()).isEqualTo("n3wp4ssw0rd");
        assertThat(actualAuthAccount.getSteamGuardToken()).isEqualTo("FGHIJ");
    }

    @Test
    void whenSetSteamAccountWithoutPassword_thenOriginalPasswordIsKept() {
        SteamAuth prevAuth = new SteamAuth(1L, "username", "hunter2", "ABCDE");
        when(steamAuthRepository.findAll()).thenReturn(List.of(prevAuth));
        SteamAuthDto newAuth = new SteamAuthDto();
        newAuth.setUsername("new_username");
        newAuth.setSteamGuardToken("FGHIJ");

        service.setAuthAccount(newAuth);

        SteamAuth actualAuthAccount = service.getAuthAccount();
        assertThat(actualAuthAccount.getUsername()).isEqualTo("new_username");
        assertThat(actualAuthAccount.getPassword()).isEqualTo("hunter2");
        assertThat(actualAuthAccount.getSteamGuardToken()).isEqualTo("FGHIJ");
    }

    @Test
    void whenClearAccount_thenRepositoryDeleteAllMethodCalled() {
        service.clearAuthAccount();

        verify(steamAuthRepository).deleteAll();
    }

    @Test
    void whenIsAuthConfiguredAndUsernameAndPasswordAreSet_thenReturnTrue() {
        SteamAuth steamAuth = new SteamAuth(1L, "username", "password", "ABCDE");
        when(steamAuthRepository.findAll()).thenReturn(List.of(steamAuth));

        boolean result = service.isAuthConfigured();

        assertThat(result).isTrue();
    }

    @Test
    void whenIsAuthConfiguredAndUsernameIsNull_thenReturnFalse() {
        SteamAuth steamAuth = new SteamAuth(1L, null, "password", "ABCDE");
        when(steamAuthRepository.findAll()).thenReturn(List.of(steamAuth));

        boolean result = service.isAuthConfigured();

        assertThat(result).isFalse();
    }

    @Test
    void whenIsAuthConfiguredAndPasswordIsNull_thenReturnFalse() {
        SteamAuth steamAuth = new SteamAuth(1L, "username", null, "ABCDE");
        when(steamAuthRepository.findAll()).thenReturn(List.of(steamAuth));

        boolean result = service.isAuthConfigured();

        assertThat(result).isFalse();
    }
}

File: backend/src/test/kotlin/cz/forgottenempire/servermanager/steamauth/SteamAuthServiceUnitTest.kt
package cz.forgottenempire.servermanager.steamauth

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.Mockito.verify
import org.mockito.Mockito.`when`
import org.mockito.junit.jupiter.MockitoExtension

@ExtendWith(MockitoExtension::class)
class SteamAuthServiceUnitTest {

    @Mock
    private lateinit var steamAuthRepository: SteamAuthRepository

    private lateinit var service: SteamAuthService

    @BeforeEach
    fun setUp() {
        service = SteamAuthService(steamAuthRepository)
    }

    @Test
    fun `when get auth account and steam auth setup, then return correct steam auth`() {
        val steamAuth = SteamAuth(1L, "username", "hunter2")
        `when`(steamAuthRepository.findAll()).thenReturn(listOf(steamAuth))

        val authAccount = service.getAuthAccount()

        assertThat(authAccount).isEqualTo(steamAuth)
    }

    @Test
    fun `when get auth account and no steam auth setup, then return empty steam auth`() {
        `when`(steamAuthRepository.findAll()).thenReturn(emptyList())

        val authAccount = service.getAuthAccount()

        assertThat(authAccount).isEqualTo(SteamAuth())
    }

    @Test
    fun `when save credentials, then username and password are persisted`() {
        val prevAuth = SteamAuth(1L, "username", "hunter2")
        `when`(steamAuthRepository.findAll()).thenReturn(listOf(prevAuth))

        service.saveCredentials("new_username", "n3wp4ssw0rd")

        val actualAuthAccount = service.getAuthAccount()
        assertThat(actualAuthAccount.username).isEqualTo("new_username")
        assertThat(actualAuthAccount.password).isEqualTo("n3wp4ssw0rd")
    }

    @Test
    fun `when clear account, then repository deleteAll method called`() {
        service.clearAuthAccount()

        verify(steamAuthRepository).deleteAll()
    }

    @Test
    fun `when is auth configured and username and password are set, then return true`() {
        val steamAuth = SteamAuth(1L, "username", "password")
        `when`(steamAuthRepository.findAll()).thenReturn(listOf(steamAuth))

        val result = service.isAuthConfigured()

        assertThat(result).isTrue()
    }

    @Test
    fun `when is auth configured and username is null, then return false`() {
        val steamAuth = SteamAuth(1L, null, "password")
        `when`(steamAuthRepository.findAll()).thenReturn(listOf(steamAuth))

        val result = service.isAuthConfigured()

        assertThat(result).isFalse()
    }

    @Test
    fun `when is auth configured and password is null, then return false`() {
        val steamAuth = SteamAuth(1L, "username", null)
        `when`(steamAuthRepository.findAll()).thenReturn(listOf(steamAuth))

        val result = service.isAuthConfigured()

        assertThat(result).isFalse()
    }
}
